Corrected small bug in unserialize (detected by Jamiras)

nanochess opened the pull request.

twinaphex commented:
Hi @nanochess , I will merge this for now.

Just out of curiosity, did you manage to talk to m4xw? He says he wants to later overwrite this particular implementation with his own which will cover all bases (this apparently leaves some spots open).

Can we out of pragmatism agree on this when this happens without either side feeling particularly done hard by? We both agree that ultimately functionality is all that matters, right?

I'm just trying to see if you guys had talked yet about this and if an agreement had been reached on a way to move forward.

nanochess commented:
Hi @twinaphex .

m4xw contacted me on Twitter by Aug/11 after I tweeted my contribution.

After reading m4xw comments, I'm sure I can add the frame video to serialization, and also some extra data about audio. These would be the only things that would be needed to restart at any point of time. I intend to do it this week.

So once implemented this, only remains the comments about structure. I think the emulator is very immature for the ctx() macro (I still want to implement/reorganize some things), and I prefer to keep my serialize/unserialize functions per module (it is a matter of personal preference because my experience).

Embarrassingly, tonight after reading your message I read again the Twitter thread, and apparently we didn't agree on who is doing the maintenance. (^^)!

This is the full text of our public Twitter exchange after I tweeted I contributed Save States to FreeIntv (Twitter handles):

`m4xwdev: Just check pending pr's next time While yours is less invasive, its also more incomplete. I planned on merging mine two weeks ago but forgot about it Might need to revert your implementation=/ I also specifically opted not to make serialize funcs, its a pita IME with other cores

nanochess: Saw your PR and the "NOT MERGE", I think the ctx() macro would make difficult supporting the code. So I asked to Blaze if I could contribute a serialization function, and got green light. Please could you elaborate on incomplete? I know very well the code as I fixed all the bugs.

m4xwdev: I already added full support for the ctx macro, in my experience it makes it easier to maintain as we have no upstream, that experience is based on 50+ cores dealing with it. It's fair if you aren't a fan of it. Calling the init funcs in unserialize and frame method is no good

m4xwdev: I also specifically allow taking a serialization at any given point in time, which yours wouldn't be able to handle

m4xwdev: Also gets rid of those nasty globals in the files, I should've moved cp and stic context to substructd tho. For memory and others I could suggest a proxy macro. The pr was up for over a month, we had all the time in the world to work on suggestions :)

nanochess: Thanks for your suggestions! Surely I can enhance it.

m4xwdev: Also:with yours you still have to update your serialize func due to not having a main context, with mine the behaviour is append to the structs, bump the ver and just handle uninitialized new variables in code This is a major improvement over most other cores, serialization wise.

m4xwdev: If there was a upstream, this would be annoying to maintain tho, i agree on that.

m4xwdev: Forgot to point out: The do-not-merge tag was just to prevent hasty merges like the one happened here ;) The implementation is/was complete since a while, just cleanup like doing a substruct for stic/cp

nanochess: Sorry about that, as I told you I asked for authorization to not duplicate efforts. Anyway I think this will make the core even better. There is a guide of how libretro handles rewind or how to test it? The current code works perfectly for save states.

m4xwdev: You just need to enable it and use the corresponding hotkeys. If theres some dynamic allocations you rely on before the first retro_run call, you might need to defer those things.

m4xwdev: Also no need to be sorry. I am more annoyed about the hasted merge than duplicated efforts, this wasnt your fault.

nanochess: Thank you. I want to make it the best possible.
`
